src/invariant/OrderBookIsNotCrossed.h

```diff
@@ -0,0 +1,99 @@
+#ifdef BUILD_TESTS
+#pragma once
+
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/Invariant.h"
+#include "ledger/LedgerHashUtils.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "xdr/Stellar-ledger-entries.h"
+#include "xdr/Stellar-ledger.h"
+
+#include <set>
+#include <unordered_map>
+
+namespace stellar
+{
+// compare two OfferEntry's by price
+struct OfferEntryCmp
+{
+    bool
+    operator()(OfferEntry const& a, OfferEntry const& b) const
+    {
+        double priceA = double(a.price.n) / double(a.price.d);
+        double priceB = double(b.price.n) / double(b.price.d);
+
+        if (priceA < priceB)
+        {
+            return true;
+        }
+        else if (priceA == priceB)
+        {
+            return a.offerID < b.offerID;
+        }
+        else
+        {
+            return false;
+        }
+    }
+};
+
+using Orders = std::set<OfferEntry, OfferEntryCmp>;
+using OrderBook = std::unordered_map<AssetPair, Orders, AssetPairHash>;
+using AssetPairSet = std::set<std::pair<Asset, Asset>>;
```

Comment:
Why not `unordered_set`?

Reply:
I have actually tried it at a few different stages of development, but so far I haven't found a point where fuzzing was clearly faster with `unordered_set`, and at times it's been slower.  There's a lot of variation on my laptop so it's hard to be sure, and I was surprised at first, but it's actually not implausible that in this case `set` might be faster:  this is a fuzzing-only invariant and when fuzzing there are only ever a very small number of orders for a given asset pair, so we might be doing one or zero comparisons typically with `set`, and with `unordered_set` we always have to do one hash.

If we ever make this invariant workable on production-size order books, then I would guess that `unordered_set` would become clearly better, but the current version of this invariant isn't.  :)

Reply:
That's a good reason.